src/unix/socket.c

```diff
@@ -0,0 +1,579 @@
+     * be persisted in the filesystem, the buffer must be empty when
+     * filesystem_add_tuple() is called, and must be filled afterwards. */
+    buffer b = allocate_buffer(sock->h, sizeof(u64));
+    table_set(s->fs_entry, sym(socket), b);
+    sysreturn ret = filesystem_add_tuple(unixaddr->sun_path, s->fs_entry);
+    if (ret) {
+        deallocate_buffer(b);
+        deallocate_tuple(s->fs_entry);
+        s->fs_entry = 0;
+        if (ret == -EEXIST) {
+            return -EADDRINUSE;
+        }
+        else {
+            return ret;
+        }
+    }
+    buffer_write_le64(b, u64_from_pointer(s));
+
+    runtime_memcpy(&s->local_addr, addr, addrlen);
+    return ret;
+}
+
+static sysreturn unixsock_listen(struct sock *sock, int backlog)
+{
+    unixsock s = (unixsock) sock;
+    if (!s->conn_q) {
+        s->conn_q = allocate_queue(sock->h, backlog);
+        if (s->conn_q == INVALID_ADDRESS) {
+            msg_err("failed to allocate connection queue\n");
+            s->conn_q = 0;
+            return -ENOMEM;
+        }
+    }
+    return 0;
+}
+
+closure_function(2, 1, sysreturn, connect_bh,
+                 unixsock, s, thread, t,
+                 u64, bqflags)
+{
+    unixsock s = bound(s);
+    thread t = bound(t);
+    sysreturn rv;
+
+    if (bqflags & BLOCKQ_ACTION_NULLIFY) {
+        rv = -EINTR;
+        goto out;
+    }
+    if (!s->connecting) {   /* the listening socket has been shut down */
+        rv = -ECONNREFUSED;
+        goto out;
+    }
+    if (!s->peer) {
+        if (s->sock.f.flags & SOCK_NONBLOCK) {
+            rv = -EINPROGRESS;
+            goto out;
+        }
+        return BLOCKQ_BLOCK_REQUIRED;
+    }
+    s->connecting = false;
+    rv = 0;
+out:
+    set_syscall_return(t, rv);
+    if (bqflags & BLOCKQ_ACTION_BLOCKED) {
+        thread_wakeup(t);
+    }
+    closure_finish();
+    return rv;
+}
+
+static sysreturn unixsock_connect(struct sock *sock, struct sockaddr *addr,
+        socklen_t addrlen)
+{
+    if (!addr) {
+        return -EFAULT;
+    }
+    unixsock s = (unixsock) sock;
+    if (s->connecting && !s->peer) {
+        return -EALREADY;
+    } else if (!s->connecting && s->peer) {
+        return -EISCONN;
+    }
+
+    struct sockaddr_un *unixaddr = (struct sockaddr_un *) addr;
+    tuple t;
+    buffer b;
+    unixsock listener, peer;
+    if (filesystem_get_tuple(unixaddr->sun_path, &t) < 0) {
+        return -ECONNREFUSED;
+    }
+    b = table_find(t, sym(socket));
+    if (!b || (buffer_length(b) != sizeof(u64))) {
+        return -ECONNREFUSED;
+    }
+    listener = pointer_from_u64(*((u64 *) buffer_ref(b, 0)));
+    assert(listener);
+    if (!s->connecting) {
+        if (!listener->conn_q || queue_full(listener->conn_q)) {
+            return -ECONNREFUSED;
+        }
+        peer = unixsock_alloc(sock->h, sock->type, 0);
+        if (!peer) {
+            return -ENOMEM;
+        }
+
+        peer->peer = s;
+        enqueue(listener->conn_q, peer);
```

Comment:
assert for enqueue here? Fullness was checked, and there should be no race with the big kernel lock, but it can't hurt...